### PR TITLE
[NLP] Supervised Fine-tuning for GPT

### PR DESCRIPTION
**Pull request by soares-f**
# What does this PR do ?

The goal of this PR is to add a supervised fine-tuning functionality for GPT.

**Collection**:  NLP - language modeling 

# Changelog 
- Created a new class similar to GPT pretraining: MegatronGPTSFTModel
- Create a new dataset class for SFT, allowing answer-only loss and the inclusion of EOS and SEP tokens
- Create a new usage example

**Pre checks**:
- [X] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [X] Did you write any new necessary tests?
- [X] Did you add or update any necessary documentation?
- [X] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [X] New Feature
- [ ] Bugfix
- [ ] Documentation

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

**Comment by soares-f** [comment]
@MaximumEntropy

File: examples/nlp/language_modeling/tuning/megatron_gpt_sft.py [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+    if cfg.model.tensor_model_parallel_size > 1 or cfg.model.pipeline_model_parallel_size > 1:
+        app_state.model_parallel_size = cfg.model.tensor_model_parallel_size * cfg.model.pipeline_model_parallel_size
+        app_state.tensor_model_parallel_size = cfg.model.tensor_model_parallel_size
+        app_state.pipeline_model_parallel_size = cfg.model.pipeline_model_parallel_size
+        (
+            app_state.tensor_model_parallel_rank,
+            app_state.pipeline_model_parallel_rank,
+            app_state.model_parallel_size,
+            app_state.data_parallel_size,
+            app_state.pipeline_model_parallel_split_rank,
```
Comment (markelsanz14) [inline examples/nlp/language_modeling/tuning/megatron_gpt_sft.py:95]: This is not set in GPT configs, I think it's only for encoder/decoder architectures. So this will crash if not set. Please use .get() so that it returns None if not found.

File: examples/nlp/language_modeling/tuning/megatron_gpt_sft.py [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+            app_state.virtual_pipeline_model_parallel_rank,
+        ) = fake_initialize_model_parallel(
+            world_size=app_state.model_parallel_size,
```
Comment (markelsanz14) [inline examples/nlp/language_modeling/tuning/megatron_gpt_sft.py:98]: This is not correct. You are doing world_size=TP*PP, but it should be world_size=TP*PP*DP.

File: examples/nlp/language_modeling/tuning/megatron_gpt_sft.py (hunk 1 above)
Comment (MaximumEntropy) [reply examples/nlp/language_modeling/tuning/megatron_gpt_sft.py:95]: The fake initialization call will just return the split rank as 0 for GPT models so this is harmless for GPT.

File: examples/nlp/language_modeling/tuning/megatron_gpt_sft.py (hunk 2 above)
Comment (MaximumEntropy) [reply examples/nlp/language_modeling/tuning/megatron_gpt_sft.py:98]: @soares-f this bit of code is probably something that is used in GPT inference code because it assumes DP=1, like @markelsanz14 said. Also I don't think you need these bits since calling `trainer.fit()` should set up most things you need. This script should look very similar to the GPT pre-training script.

**Review by XuesongYang — CHANGES_REQUESTED** [review]
please rebase to the latest main branch, otherwise, the unnecessary changes for TTS related codes would crash. Thanks.
